src/man/man8/samgetmap.8
'\" t
.\" $Revision: 1.17 $
.ds ]W Sun Microsystems
'\" !tbl | mmdoc
.\" SAM-QFS_notice_begin
.\"
.\" CDDL HEADER START
.\"
.\" The contents of this file are subject to the terms of the
.\" Common Development and Distribution License (the "License").
.\" You may not use this file except in compliance with the License.
.\"
.\" You can obtain a copy of the license at pkg/OPENSOLARIS.LICENSE
.\" or https://illumos.org/license/CDDL.
.\" See the License for the specific language governing permissions
.\" and limitations under the License.
.\"
.\" When distributing Covered Code, include this CDDL HEADER in each
.\" file and include the License file at pkg/OPENSOLARIS.LICENSE.
.\" If applicable, add the following below this CDDL HEADER, with the
.\" fields enclosed by brackets "[]" replaced with your own identifying
.\" information: Portions Copyright [yyyy] [name of copyright owner]
.\"
.\" CDDL HEADER END
.\"
.\" Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
.\" Use is subject to license terms.
.\"
.\" SAM-QFS_notice_end
.na
.nh
.TH samgetmap 8 "26 Feb 2001"
.SH NAME
samgetmap \- Obtains disk file storage information
.SH SYNOPSIS
\fB/opt/SUNWsamfs/tools/samgetmap\fR
\%[\fB\-a \fIallocsize\fR]
\%[\fB\-c\fR]
\%[\fB\-f\fR]
\%[\fB\-h\fR]
\%[\fB\-l \fIsetlen\fR]
\%[\fB\-m \fIminalloc\fR]
\%[\fB\-n \fInbytes\fR]
\%[\fB\-s \fIstartaddr\fR]
\%[\fB\-u\fR]
\%[\fB\-w\fR]
\%[\fB\-M\fR]
\%[\fB\-U\fR]
\%[\fB\-V\fR]
\fIfile\fR
.SH AVAILABILITY
\fBSun Microsystems Internal\fR
.SH DESCRIPTION
The \fBsamgetmap\fR command
provides a test interface to the SANergy File Map API routines.
The \fBsamgetmap\fR command must be run as root.
.SH OPTIONS
This command accepts the following options:
.TP 10
\fB\-a \fIallocsize\fR
Requests allocation of \fIallocsize\fR bytes of storage to \fIfile\fR
through the \fBFS_SetFileSizes\fR request.
The \fIallocsize\fR must be an integer in the following range:
.sp
0 \(<= \fIallocsize\fR < 2**31
.sp
For more information, see the OPERATIONS section
of this man page.
.TP
\fB\-c\fR
Requests a canonical map.
\fBFS_M_FLAG_CANONICAL\fR is set for the \fBFS_GetLockedMap\fR request.
For more information, see the OPERATIONS section
of this man page.
.TP
\fB\-f\fR
Requests a non-sparse allocation map.
\fBFS_M_FLAG_NO_HOLE\fR is set for the \fBFS_GetLockedMap\fR request.
For more information, see the OPERATIONS section
of this man page.
.TP
\fB\-h\fR
Causes \fBsamgetmap\fR to write a short usage message to \fBstdout\fR.
.TP
\fB\-l \fIsetlen\fR
Sets the file length to \fIsetlen\fR bytes.
.TP
\fB\-m \fIminalloc\fR
Uses \fBFS_SetFileSizes\fR to
request the allocation of \fIminalloc\fR bytes of storage for \fIfile\fR.
\fIminalloc\fR must be an integer in the following range:
.sp
-1 \(<= \fIminalloc\fR \(<= \fIallocsize\fR
.sp
If \fIminalloc\fR is within this range and the allocation request
partially succeeds, such that \fIminalloc\fR or more bytes are
allocated but fewer than \fIallocsize\fR bytes are allocated,
a unique error code is returned to indicate partial success.
For more information, see the OPERATIONS section of this man page.
.sp
\fIsamgetmap\fR's behavior is undefined if \fIminalloc\fR
is less than \fB-1\fR or greater than \fIallocsize\fR.
For more information on \fIallocsize\fR, see the \fB\-a\fR option.
.TP
\fB\-n \fInbytes\fR
Specifies that the allocation map returned by
the \fBFS_GetLockedMap\fR call return information about the
location of at least \fInbytes\fR of data.
For more information, see the OPERATIONS section
of this man page.
.TP
\fB\-s \fIstartaddr\fR
Requests that the allocation map returned by the \fBFS_GetLockedMap\fR
request return the information about \fIfile\fR's storage beginning at
byte \fIstartaddr\fR in the file.
For more information, see the OPERATIONS section
of this man page.
.TP
\fB\-u\fR
Sets \fBFS_M_FLAG_UNLOCKED\fR in the call to \fBFS_GetLockedMap\fR.
\fIfile\fR is not locked by the \fBFS_GetLockedMap\fR call.
For more information, see the OPERATIONS section
of this man page.
.TP
\fB\-w\fR
Sets \fBFS_M_FLAG_WAIT\fR for the call to \fBFS_GetLockedMap\fR.
If the file is online, this option has no effect.
If the file is offline, the file is staged in before
the file's map is returned.
For more information, see the OPERATIONS section
of this man page.
.TP
\fB\-M\fR
Prevents the call to \fBFS_GetLockedMap\fR from being made.
For more information, see the OPERATIONS section
of this man page.
.TP
\fB\-U\fR
Prevents the call to \fBFS_UnlockMap\fR from being made, leaving
the file locked (unless the \fB-u\fR option is also specified).
For more information, see the OPERATIONS section
of this man page.
.sp
NOTE:  Use of this option makes it impossible to unmount the file
system that
\fIfile\fR resides in until the file is somehow unlocked (see also
\fBsamunhold\fR(8) to remove all file locks (holds) on a file system).
Specifying the \fB\-u\fR and \fB\-U\fR options together on the
command line insures that the
file is unlocked upon exit from \fBsamgetmap\fR.
.TP
\fB-V\fR
Report the value returned by the FS_GetVersion(), added to the
SANergy 2.2 API, which indicates the library revision value.
Present valid values are 220 (for SANergy 2.2),
and 230 (for SANergy 2.3/3.1).
.TP
\fIfile\fR
Specifies the file for which disk storage information is requested.
.PP
The \fBsamgetmap\fR command
should not be executed if SANergy File Sharing
is running on the file system.
Specifically, the administrator
should ensure that the following conditions are true:
.TP 2
\(bu
There are no SANergy applications running on any
client, possibly including the server itself.
.TP
\(bu
The file system in question is not fused on any SANergy clients.
.PP
.SH OPERATIONS
The \fBsamgetmap\fR program operates as follows:
.TP 3
1.
The program calls the \fBAFS_GetCookies\fR routine
to obtain two cookies, one for the file system
that \fIfile\fR resides on (a volume cookie)
and the other for \fIfile\fR itself (a file cookie).
.TP
2.
The program calls \fBFS_GetLockedMap\fR, using the cookies
obtained by \fBFS_GetCookies\fR.
This call is not made if the \fB-M\fR option is present.
The parameters to this call can be specified or modified by the
command line options \fB\-s\fR, \fB\-n\fR, \fB\-c\fR,
\fB\-f\fR, \fB\-u\fR, and \fB\-w\fR.
.TP
3.
The program calls \fBFS_SetFileSizes\fR.
The parameters to this call can be specified or modified by the
command line options \fB\-a\fR, \fB\-l\fR, and \fB\-m\fR.
.TP
4.
The program calls \fBFS_UnlockMap\fR.
This call is not made if the \fB\-U\fR option is present on
the command line.
.PP
The result of each call is written in a message to \fBstdout\fR.
.SH EXAMPLES
The following example shows how to use \fBsamgetmap\fR:
.PP
.nf
.ft CO
ceres# samgetmap /qfs1/foo
AFS_GetCookies("/qfs1/foo", &vc, &fc) = 0 (FS_E_SUCCESS (OK))

Volcookie:
          0  0  0 20 73 56 6d 43          0  b  0  0 3a 94 44 4d
          0  1  5  6  0  0  0  0          0  0  3  0  2 ed a0  0

Filecookie:
          0  0  0 10 73 46 6d 43          0  0  0  7  0  0  0  1
          0  0  0  0  0  0  0  0          0  0  0  0  0  0  0  0

FS_GetLockedMap(&vc, &fc, 0, -1, 0, ffbaf448, ffbaf340)  = 0 (FS_E_SUCCESS
(OK))
returned buflen = 88
msgLen       = 88
vendorStatus = 0
fileSize     = 16384000
allocation   = 16416768
nExtents     = 1
extentType   = 2        (SIMPLE)

sExtent[0]
        volumeOrdinal = 1
        blockOffset   = 20
        nBlocks = 7d40

FS_SetFileSizes(&vc, &fc, 0, ffffffffffffffff, ffffffffffffffff, ffffffffff
ffffff)  = 0 (FS_E_SUCCESS (OK))
FS_UnlockMap(&vc, &fc) = 0 (FS_E_SUCCESS (OK))
.ft
.fi
.SH NOTES
The \fBsamgetmap\fR command typically issues holds (locks) to \fIfile\fR.
This can interfere with the operation of SANergy File Sharing,
possibly causing file system corruption.
.SH SEE ALSO
\fBsamgetvol\fR(8),
\fBsamunhold\fR(8).



src/fs/tools/san/samgetmap.c
/*
 * samgetmap.c - dump file allocation ("map") information
 */

/*
 *    SAM-QFS_notice_begin
 *
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at pkg/OPENSOLARIS.LICENSE
 * or https://illumos.org/license/CDDL.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at pkg/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 * Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 *
 *    SAM-QFS_notice_end
 */


#pragma ident "$Revision: 1.22 $"

#include <stdio.h>
#include <stdlib.h>
#include <stddef.h>
#include <string.h>
#include "sys/types.h"
#include "samsanergy/fsmdc.h"
#include "samsanergy/fsmdcsam.h"

static char *extent_format[] = {
	"COMPACT",
	"VERBOSE",
	"SIMPLE",
	"CANONICAL"
};

static char *extent_validity[] = {
	"DIRECT",
	"HOLE",
	"RESERVED1",
	"RESERVED2",
	"NO ACCESS",
	"NOT PRESENT",
	"READ VALID",
	"READ ONLY"
};

static void
dumpmap(FSMAPINFO *mp)
{
	int i;

	printf("fileSize     = %lld\t[%#llx]\n", mp->fileSize, mp->fileSize);
	printf("allocation   = %lld\t[%#llx]\n",
	    mp->allocation, mp->allocation);
	printf("msgLen       = %ld\t[%#lx]\n",
	    (long)mp->msgLen, (long)mp->msgLen);
	printf("vendorStatus = %ld\t[%#lx]\n",
	    (long)mp->vendorStatus, (long)mp->vendorStatus);
	printf("nExtents     = %ld\t[%#lx]\n",
	    (long)mp->nExtents, (long)mp->nExtents);
	printf("extentType   = %ld", (long)mp->extentType);
	if (mp->extentType >= 0 && mp->extentType < 4) {
		printf("\t(%s)\n", extent_format[mp->extentType]);
	} else {
		printf("\t(INVALID EXTENT FORMAT)\n");
	}
	printf("fileOffset   = %lld\t[%#llx]\n",
	    (long long)mp->fileOffset,
	    (long long)mp->fileOffset);
	printf("fileBytes    = %lld\t[%#llx]\n",
	    (long long)mp->fileBytes, (long long)mp->fileBytes);
	printf("\n");

	switch (mp->extentType) {
	case FS_EXTENT_COMPACT:
		for (i = 0; i < mp->nExtents; i++) {
			printf("cExtent[%d]\n", i);
			printf("\tblockOffset = %llx\t",
			    (long long)mp->cExtent[i].blockOffset);
			printf("\tnBlocks     = %llx\n\n",
			    (long long)mp->cExtent[i].nBlocks);
		}
		break;
	case FS_EXTENT_SIMPLE:
		for (i = 0; i < mp->nExtents; i++) {
			printf("sExtent[%d]\n", i);
			printf("\tvolumeOrdinal = %lld\t[%#llx]\n",
			    (long long)mp->sExtent[i].volumeOrdinal,
			    (long long)mp->sExtent[i].volumeOrdinal);
			printf("\tblockOffset   = %lld\t[%#llx]\n",
			    (long long)mp->sExtent[i].blockOffset,
			    (long long)mp->sExtent[i].blockOffset);
			printf("\tnBlocks       = %lld\t[%#llx]\n\n",
			    (long long)mp->sExtent[i].nBlocks,
			    (long long)mp->sExtent[i].nBlocks);
		}
		break;
	case FS_EXTENT_VERBOSE:
	case FS_EXTENT_CANONICAL:
		for (i = 0; i < mp->nExtents; i++) {
			printf("vExtent[%d]\n", i);
			printf("\tvolumeOrdinal     = %ld\t[%#lx]\n",
			    (long)mp->vExtent[i].volumeOrdinal,
			    (long)mp->vExtent[i].volumeOrdinal);
			printf("\tlogicalByteOffset = %lld\t[%#llx]\n",
			    (long long)mp->vExtent[i].logicalByteOffset,
			    (long long)mp->vExtent[i].logicalByteOffset);
			printf("\tphysicalByteOffset= %lld\t[%#llx]\n",
			    (long long)mp->vExtent[i].physicalByteOffset,
			    (long long)mp->vExtent[i].physicalByteOffset);
			printf("\tnBytes            = %lld\t[%#llx]\n",
			    (long long)mp->vExtent[i].nBytes,
			    (long long)mp->vExtent[i].nBytes);
			printf("\tvalidity          = %ld\t[%#lx]",
			    (long)mp->vExtent[i].validity,
			    (long)mp->vExtent[i].validity);
			if (mp->vExtent[i].validity >= 0 &&
			    mp->vExtent[i].validity < 8) {
				printf("\t(%s)\n\n",
				    extent_validity[mp->vExtent[i].validity]);
			} else {
				printf("\t(INVALID EXTENT TYPE)\n\n");
			}
		}
		break;
	default:
		printf("BAD EXTENT FORMAT (%d).\n\n", mp->extentType);
	}
}

static char *
mapretcode(int n, char *buf)
{
	switch (n) {
	case FS_E_SUCCESS:		/* Successful normal completion */
		return ("FS_E_SUCCESS (OK)");

	case FS_E_MISC_ERROR:		/* Vendor-specific error */
		return ("FS_E_MISC_ERROR (GENERIC ERROR)");

	case FS_E_PARTIAL:		/* Success, but partial information */
		return ("FS_E_PARTIAL (OK; BUT BUFFER TOO SMALL)");

	case FS_E_WRONG_FS:		/* Not owned by this FSMDC */
		return ("FS_E_WRONG_FS (NOT SAM-FS/QFS FILE)");

	case FS_E_BAD_PATH:		/* Illegal or bad path or share name */
		return ("FS_E_BAD_PATH (NO FILE)");

	case FS_E_BAD_COOKIE:		/* Bad volume cookie */
		return ("FS_E_BAD_COOKIE (BAD OR STALE COOKIE)");

	case FS_E_OUT_OF_MEMORY:	/* Out of memory */
		return ("FS_E_OUT_OF_MEMORY (CAN'T ALLOCATE MEMORY)");

	case FS_E_INVALID_MSG_LENGTH:	/* Bad message length */
		return ("FS_E_INVALID_MSG_LENGTH (?)");

	case FS_E_INVALID_ENDIAN:	/* Unrecognized "endian"ness? */
		return ("FS_E_INVALID_ENDIAN (?)");

	case FS_E_UNKNOWN_FUNCTION:	/* Unrecognized function? */
		return ("FS_E_UNKNOWN_FUNCTION (?)");

	case FS_E_UNKNOWN_SHARE:	/* ? */
		return ("FS_E_UNKNOWN_SHARE (?)");

	case FS_E_UNKNOWN_FILE:		/* stale or bad cookie */
		return ("FS_E_UNKNOWN_FILE (BAD OR STALE COOKIE #2)");

	case FS_E_UNKNOWN_STREAM:	/* unknown file stream (NT) */
		return ("FS_E_UNKNOWN_STREAM (UNKNOWN FILE STREAM -- "
		    "NT ERROR?)");

	case FS_E_NO_DIRECT:		/* valid file, but no direct access */
		return ("FS_E_NO_DIRECT (NO DIRECT ACCESS TO FILE "
		    "(SPECIAL? DIR?))");

	case FS_E_PRIVILEGE:		/* caller has insufficient privilege */
		return ("FS_E_PRIVILEGE (INSUFFICIENT PRIVILEGE)");

	case FS_E_BAD_FILE:		/* Bad file or volume cookie */
		return ("FS_E_BAD_FILE (BAD OR STALE COOKIE #3)");

	case FS_E_SUBMAP:		/* Success, but submap returned */
		return ("FS_E_SUBMAP (FILE NOT FULLY RESIDENT?)");

	case FS_E_IVPARAM:		/* Start value not aligned */
		return ("FS_E_IVPARAM (FILE START OFFSET NOT ALIGNED?)");

	case FS_E_NOT_LOCKED:		/* File wasn't locked to begin with */
		return ("FS_E_NOT_LOCKED (UNLOCK OF UNLOCKED FILE)");

	case FS_E_SHORT_ALLOC:		/* Success, but not fully allocated */
		return ("FS_E_SHORT_ALLOC (FILE ALLOCATION PARTIALLY "
		    "SUCCESSFUL)");

	case FS_E_LICENSE_ERR:		/* License problem */
		return ("FS_E_LICENSE_ERR (?)");

	case FS_E_LICENSE_DUP:		/* duplicated license */
		return ("FS_E_LICENSE_DUP (?)");

	case FS_E_LICENSE_INV:		/* invalid license */
		return ("FS_E_LICENSE_INV (?)");

	}
	sprintf(buf, "UNKNOWN ERROR (%#x)", n);
	return (buf);
}


static char *usage =
	"Usage:\n\tsamgetmap\n"
	"\t	[-h]			-- print help message\n"
	"\t	[-a allocsize]		-- preallocate to allocsize\n"
	"\t	[-l setlen]		-- set file size to setlen\n"
	"\t	[-m minalloc]		-- allocate at least minalloc\n"
	"\t	[-c]			-- request canonical map\n"
	"\t	[-f]			-- fill allocation holes in map\n"
	"\t	[-u]			-- don't lock file\n"
	"\t	[-w]			-- wait (for staging?)\n"
	"\t	[-r]			-- retrieve (initiate stage)\n"
	"\t	[-s filestart]		-- start file map at filestart\n"
	"\t	[-n filebytes]		-- get map for at least filebytes\n"
	"\t	[-M]			-- don't get map\n"
	"\t	[-U]			-- don't unlock file\n"
	"\t	file\n\n";

int Mflag = 0;
int Uflag = 0;

int
main(int ac, char *av[])
{
	unsigned long long buf[32*1024];
	char msgbuf[256];
	FSULONG flags = 0, buflen = sizeof (buf);
	FS64LONG filesize = (FS64LONG)-1,
	    allocmin = (FS64LONG)-1,
	    allocsize = (FS64LONG)-1;
	FS64LONG fstart = 0, flen = (FS64LONG)-1;
	FSMAPINFO *smp = (FSMAPINFO *)buf;
	FSVOLCOOKIE vc;
	FSFILECOOKIE fc;
	int c, i;
	FSLONG r;
	extern int errno;

	while ((c = getopt(ac, av, "a:cfhl:m:Mn:rs:uUVw")) != EOF) {
		switch (c) {
		case 'a':		/* set allocated size */
			allocsize = atoll(optarg);
			break;
		case 'c':
			flags |= FS_M_FLAG_CANONICAL;
			break;
		case 'l':		/* set file len */
			filesize = atoll(optarg);
			break;
		case 'm':		/* min alloc size */
			allocmin = atoll(optarg);
			break;

		case 'f':		/* fill any allocation holes */
			flags |= FS_M_FLAG_NO_HOLE;
			break;
		case 'u':		/* don't lock file */
			flags |= FS_M_FLAG_UNLOCKED;
			break;
		case 'r':
			flags |= FS_M_FLAG_RETRIEVE;
			break;
		case 'w':		/* WAIT for entire submap */
			flags |= FS_M_FLAG_WAIT;
			break;

		case 'M':		/* don't get map */
			Mflag = 1;
			break;
		case 'U':		/* don't unlock file */
			Uflag = 1;
			break;

		case 's':		/* file map start offset */
			fstart = atoll(optarg);
			break;
		case 'n':		/* file map len */
			flen = atoll(optarg);
			break;

		case 'V':
			printf("Version (FS_GetVersion) = %ld\n",
			    (long)FS_GetVersion());
			break;

		case 'h':
		default:
			printf(usage);
			exit(c != 'h');
		}
	}
	if (optind+1 != ac) {
		printf(usage);
		exit(1);
	}
	errno = 0;
	r = AFS_GetCookies(av[optind], &vc, &fc);
	printf("AFS_GetCookies(\"%s\", &vc, &fc) = %#lx (%s)\n",
	    av[optind], (long)r, mapretcode(r, msgbuf));
	printf("\nVolcookie:\n");
	for (i = 0; i < sizeof (vc); i++) {
		if ((i%8) == 0)
			printf("\t");
		printf(" %2x", vc.xprivate[i]&0xff);
		if ((i%16) == 15)
			printf("\n");
	}
	printf("\nFilecookie:\n");
	for (i = 0; i < sizeof (fc); i++) {
		if ((i%8) == 0)
			printf("\t");
		printf(" %2x", fc.xprivate[i]&0xff);
		if ((i%16) == 15)
			printf("\n");
	}
	printf("\n");

	if (!Mflag) {
		r = FS_GetLockedMap(&vc, &fc,
		    fstart, flen,
		    flags,
		    smp, &buflen);
		printf("FS_GetLockedMap(&vc, &fc, %lld, %lld, %lx, %llx, %llx)"
		    "  = %#lx (%s)\n",
		    fstart, flen, (long)flags,
		    (long long)buf, (long long)&buflen,
		    (long)r, mapretcode(r, msgbuf));
		if (r != FS_E_SUCCESS) {
			sam_fsuni2mb(msgbuf, smp->vendorError,
			    sizeof (buf)-offsetof(FSMAPINFO, vendorError)),
			    printf("vendorStatus=%d, vendorError='%s', "
			    "errno='%s' (%d)\n",
			    smp->vendorStatus, msgbuf,
			    strerror(errno), (int)errno);
		}
		printf("returned buflen = %d\n", (int)buflen);
		dumpmap(smp);
	}

	r = FS_SetFileSizes(&vc, &fc, 0, filesize, allocsize, allocmin);
	printf("FS_SetFileSizes(&vc, &fc, 0, %llx, %llx, %llx)  = %#lx (%s)\n",
	    filesize, allocsize, allocmin,
	    (long)r, mapretcode(r, msgbuf));

	if (!Uflag) {
		r = FS_UnlockMap(&vc, &fc);
		printf("FS_UnlockMap(&vc, &fc) = %#lx (%s)\n",
		    (long)r, mapretcode(r, msgbuf));
	}

	return (0);
}
